sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignature.java

```diff
@@ -113,12 +117,148 @@ static Builder builder() {
     Method targetMethod();
   }
 
-  /** A type of optional parameter of the {@link DoFn.ProcessElement} method. */
-  public enum Parameter {
-    BOUNDED_WINDOW,
-    INPUT_PROVIDER,
-    OUTPUT_RECEIVER,
-    RESTRICTION_TRACKER
+  /** A descriptor for an optional parameter of the {@link DoFn.ProcessElement} method. */
+  public abstract static class Parameter {
+
+    // Private as no extensions other than those nested here are permitted
+    private Parameter() {}
+
+    /**
+     * Performs case analysis on this {@link Parameter}, processing it with the appropriate
+     * {@link Cases#dispatch} case of the provided {@link Cases} object.
+     */
+    public <ResultT> ResultT match(Cases<ResultT> cases) {
+      // This could be done with reflection, but since the number of cases is small and known,
+      // they are simply inlined.
+      if (this instanceof BoundedWindowParameter) {
+        return cases.dispatch((BoundedWindowParameter) this);
+      } else if (this instanceof RestrictionTrackerParameter) {
+        return cases.dispatch((RestrictionTrackerParameter) this);
+      } else if (this instanceof InputProviderParameter) {
+        return cases.dispatch((InputProviderParameter) this);
+      }
+      if (this instanceof OutputReceiverParameter) {
+        return cases.dispatch((OutputReceiverParameter) this);
+      }
+      if (this instanceof StateParameter) {
+        return cases.dispatch((StateParameter) this);
+      } else {
+        throw new IllegalStateException(
+            String.format("Attempt to case match on unknown %s subclass %s",
+                Parameter.class.getCanonicalName(), this.getClass().getCanonicalName()));
+      }
+    }
+
+    /**
+     * A visitor for destructuring a {@link Parameter}.
```

Comment:
drop the "visitor". `An interface for destructuring {@link Parameter Parameters}.`